### ISSUE TITLE
Recover targets from the distant past to avoid rerunning commands

### ISSUE THREAD

**Issue by wlandau** [ropensci/drake#945]
## Prework

- [x] Read and abide by `drake`'s [code of conduct](https://github.com/ropensci/drake/blob/master/CODE_OF_CONDUCT.md).
- [x] Search for duplicates among the [existing issues](https://github.com/ropensci/drake/issues), both open and closed.

## Background

Let's start with a simple plan.

```r
library(drake)
library(tidyverse)

plan <- drake_plan(
  x = iris %>%
    group_by(Species) %>%
    summarize_all(mean)
)

make(plan)
# target x
```

As we work on the project, we make changes. For example:

```r
library(drake)
library(tidyverse)

plan <- drake_plan(
  x = iris %>%
    group_by(Species) %>%
    summarize_all(sum) # changed from `mean` to `sum`
)

make(plan)
# target x
```

## Proposal

Suppose we go back to our original plan, e.g. using `mean` instead of `sum`. The current `drake` reruns command and builds the target. But unless we ran `clean(garbage_collection = TRUE)`, that old data is still in the cache. What if we restored the old data instead of rerunning the command?

```r
library(drake)
library(tidyverse)

plan <- drake_plan(
  x = iris %>%
    group_by(Species) %>%
    summarize_all(mean)
)

make(plan)
# target x
# restore x
```

**Comment by wlandau** [comment]
We need to recover the metadata hash from the old build. Could be tricky.

**Comment by wlandau** [comment]
## Good news

I think this is possible. And for the implementation I have in mind, the target's name will not matter. In fact, `rename_targets()` should no longer be necessary (#936). Simply changing the name of the target should tell `drake` to salvage old data, not compute new data.

## Bad news

The implementation will inevitably create more tiny files in the cache. If you are worried about maxing out the file quota on your machine, this could be a problem. Maybe we should have a new logical `recoverable` argument in `make()` and `drake_config()`.

**Comment by wlandau** [comment]
And, unavoidably, renaming a target will always invalidate downstream targets because of the way `drake` computes dependency hashes.

**Comment by wlandau** [comment]
Another obstacle: seeds. A target's seed is determined by the global seed and the target name. This is also a problem for the existing `rename_targets()`. Can we

1. Find a new method to come up with a target's seed?
2. Roll out the new method without breaking existing workflows?

**Comment by wlandau** [comment]
Nope. I really think the RNG seed needs to be determined by the target name. I do not see another way to generate seeds reproducibly. So in order to restore targets, the user needs to restore the original target names.

But we can work with that. Users should be putting their projects under version control anyway. If we allow target recovery with strict names, `drake` will listen to `git reset`s.

**Comment by wlandau** [comment]
Hmm...what if we also allow a custom `seed` column in the plan? Then, renaming will indeed happen automatically. We just need to tell folks to set their own seeds!

**Comment by wlandau** [comment]
Got started today in the [`945` branch](https://github.com/ropensci/drake/tree/945). It needs more testing, but it seems to be working okay so far. Issues:

1. For a test workflow with 10000 empty targets, we spend ~7.5% of our runtime saving the metadata.
2. To avoid loading the data unnecessarily, the implementation requires `config$cache$driver$set_hash()`, which is not part of the common `storr` API. Fortunately, it is part of the drivers we care about most of the time: RDS, DBI, and environment. We can simply warn users that some features may be disabled for fringe drivers. If we do that, we could even use `set_hash()` to speed up (1), as well as `set_progress()`.

**Comment by wlandau** [comment]
Main issues I see with data recovery:

1. Speed.
2. Reproducibility. In [the branch](https://github.com/ropensci/drake/tree/945), old target values are subject to recovery no matter how far back they go. We risk salvaging data produced from a completely different package environment that no longer exists.

I think we had better hold off on this for a while. I still think data recovery, particularly automated data recovery, would be a game changer. But I am less enthused than before about the first candidate implementation.

**Comment by wlandau** [comment]
Perhaps it is enough to label this feature as experimental and disable recovery by default.

**Comment by brendanf** [comment]
> And, unavoidably, renaming a target will always invalidate downstream targets because of the way `drake` computes dependency hashes.

That's too bad.  It would be useful for `drake` to be able to reuse the cached value for a target with a different name but identical command.  Two situations where this could come up are:

```r
library(drake)
A <- 1
plan <- drake_plan(
  B = very_slow_function(A)
)
make(plan)

# okay, now I know that my plan works, time to apply it to more cases
A <- 1:5
plan2 <- drake_plan(
  B = target(
     very_slow_function(A_),
     transform = map(A_ = !!A)
  )
)

# this will have to execute very_slow_function(1) again,
# even though command of B_1 is identical to B

make(plan2) 
```

and

```r
library(drake)

### define some data

plan <- drake_plan(
  # always makes same clusters for same input, but ordering is nondetereministic
  # e.g. uses internal parallelism and results are build up depending on which
  # subprocess finishes first
  clust <- make_clusters(data, n = 5),
  oneclust <- target(
    clust,
    transform = split(clust, slices = 5)
  ),
  analysis = target(
    very_slow_function(oneclust),
    transform = map(oneclust, .id = FALSE)
  )
)
```

both of these cases could be dealt with by using
```
library(memoise)
very_slow_function <- memoise(very_slow_function, cache = cache_filesystem(here(".cache")))
```
but then we would be caching the same results in two different file system caches, which seems inefficient.

**Comment by wlandau** [comment]
Implemented in #952. Still experimental and turned off by default.
